### PR TITLE
Reveal test vectors for PegIn and PegOutRequest

### PR DESCRIPTION
**Pull request by netrome**
Initial stab at adding commit-reveal test vectors for peg-in and peg-out-request (nowadays also known as deposit reveal and withdrawal reveal).

This PR contains some code duplication between the OP_RETURN and commit-reveal implementations. I chose to postpone cleaning this up to when https://github.com/Trust-Machines/core-eng/pull/339 is ready. At that point we should make a larger refactor to reuse the commit-reveal library for more proper test vectors.

### CODE REVIEW

**Review by jo-tm — APPROVED** [review]
thanks!

**Review by stjepangolemac — APPROVED** [review]

File: test-vectors/src/peg_in.rs [hunk 1]
```diff
@@ -15,32 +19,103 @@ pub fn generate_peg_in_test_vector() -> Transaction {
     let input = vec![input];
 
     let stacks_address = [1u8; 20];
-    let stacks_testnet_singlesig_address_version = [26];
+    let stacks_testnet_singlesig_address_version = 26;
```
[the remark this answers is not in the document]
Comment (netrome) [reply test-vectors/src/peg_in.rs]: Right, I copied these constants now. I don't want to drag in stacks as a dependency here for just these constants.

File: test-vectors/src/peg_in.rs [hunk 2]
```diff
@@ -15,32 +19,103 @@ pub fn generate_peg_in_test_vector() -> Transaction {
...
+pub fn generate_peg_in_reveal_test_vector() -> Transaction {
+    let mut input = TxIn {
+        previous_output: OutPoint::null(),
+        script_sig: Script::empty().into(),
+        sequence: Sequence::MAX,
+        witness: Witness::new(),
+    };
+
+    let stacks_address = [1u8; 20];
+    let stacks_testnet_singlesig_address_version = 26;
```
[the remark this answers is not in the document]
Comment (netrome) [reply test-vectors/src/peg_in.rs]: Updated :heavy_check_mark:

**Comment by netrome** [comment]
Some test vectors generated by this code (using legacy terminology):

**Peg in OP_RETURN**
```
02000000010000000000000000000000000000000000000000000000000000000000000000ffffffff00ffffffff020000000000000000516a4c4e69643c061a010101010101010101010101010101010101010116736274632d72656365697665722d636f6e74726163740000000000000000000000000000000000000000000000000000000000003905000000000000225120040404040404040404040404040404040404040404040404040404040404040400000000
```

**Peg in OP_DROP**
```
020000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff00ffffffff020000000000000000056a036964773905000000000000225120040404040404040404040404040404040404040404040404040404040404040402764c4e3c061a010101010101010101010101010101010101010116736274632d72656365697665722d636f6e747261637400000000000000000000000000000000000000000000000000000000000000007576a920040404040404040404040404040404040404040404040404040404040404040487ad613c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c00000000
```

**Peg out request OP_RETURN**
```
02000000010000000000000000000000000000000000000000000000000000000000000000ffffffff00ffff0000030000000000000000516a4c4e69643e00000000000005390135af51a3d143749a43da266fd986fbdd7b1a1c858669ae10ee63075d15febce208bde8a35eee0deba178e6c1ae73067da918cb437db04077084960b0d78719c0000015000000000000001976a914050505050505050505050505050505050505050587ad2a00000000000000225120040404040404040404040404040404040404040404040404040404040404040400000000
```

**Peg out request OP_DROP**
```
020000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff00ffffffff030000000000000000056a0369647715000000000000001976a914050505050505050505050505050505050505050587ad2a00000000000000225120040404040404040404040404040404040404040404040404040404040404040402764c4e3e00000000000005390135af51a3d143749a43da266fd986fbdd7b1a1c858669ae10ee63075d15febce208bde8a35eee0deba178e6c1ae73067da918cb437db04077084960b0d78719c0000000007576a920040404040404040404040404040404040404040404040404040404040404040487ad613c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c3c00000000
```

**Peg handoff OP_RETURN**
```
02000000010000000000000000000000000000000000000000000000000000000000000000ffffffff00ffff0000020000000000000000536a036964480800000000000000434400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000225120040404040404040404040404040404040404040404040404040404040404040400000000
```

FYI @MarvinJanssen @setzeus @jo-tm - I hope this is useful for you. Let me know if anything looks fishy and I'll update.

**Comment by MarvinJanssen** [comment]
Three questions @netrome:

1. Is the OP_RETURN peg in back?
2. First time I'm seeing an OP_RETURN peg out, I thought we were not doing this?
3. Is peg handoff confirmed to use OP_RETURN and no longer OP_DROP?

**Comment by igorsyl** [comment]
> Three questions @netrome:
> 
> 1. Is the OP_RETURN peg in back?
> 2. First time I'm seeing an OP_RETURN peg out, I thought we were not doing this?
> 3. Is peg handoff confirmed to use OP_RETURN and no longer OP_DROP?

We're using OP_RETURN for sBTC Alpha only. For Mini and Production, it's OP_DROP only.

**Comment by netrome** [comment]
> Three questions @netrome:
> 
>     1. Is the OP_RETURN peg in back?
> 
>     2. First time I'm seeing an OP_RETURN peg out, I thought we were not doing this?
> 
>     3. Is peg handoff confirmed to use OP_RETURN and no longer OP_DROP?

1. OP_RETURN is planned for alpha, but not mini or sBTC 1.0. I anticipate that we'll revive OP_RETURN again for a subsequent sBTC release since it is the most efficient flow (saves btc fees) for anyone using a wallet supporting it.
2. For alpha, everything is OP_RETURN including peg out. Moreover, I figured better to add too many examples rather than too few. From the protocol side of things OP_DROP and OP_RETURN are easily interchangeable formats, while I know they differ a lot in the UX and from the Clarity perspective.
3. Afaik OP_DROP was decided for peg-in and peg-out. I know there has been discussions to apply it for handoff too, but I'm not sure if any decision has been made. I'm not aware of any good reason to use OP_DROP for the peg handoff, since the transaction is not originating from a custodial wallet. If we need reclaimability, we can add that on OP_RETURN just as easily as OP_DROP. However, if you want to go with OP_DROP here for any particular reason I can add a test vector for it.

**Comment by MarvinJanssen** [comment]
On point 3: one possible upside when using `OP_DROP` for the handoff is that it proves that the new signers can successfully sign a transaction. If not, then the current signers can recover the peg wallet.

**Comment by netrome** [comment]
> On point 3: one possible upside when using `OP_DROP` for the handoff is that it proves that the new signers can successfully sign a transaction. If not, then the current signers can recover the peg wallet.

True, although I'm not sure how much extra confidence that gives us from the signers as opposed to OP_RETURN where they still have to prove they can successfully run DKG to generate a public key. But sure, proving that they can sign a transaction is a step-up from that.

**Pull request merged**